File: qt/gen_qbuffer.go
package qt

import (
	"unsafe"
)

type QBuffer struct {
	h          uintptr
	isSubclass bool
}

// NewQBuffer constructs a new QBuffer object.
func NewQBuffer() *QBuffer {
	ret := newQBuffer(QBuffer_new())
	ret.isSubclass = true
	return ret
}

// NewQBuffer2 constructs a new QBuffer object.
func NewQBuffer2(parent *QObject) *QBuffer {
	ret := newQBuffer(QBuffer_new2(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QBuffer) MetaObject() *QMetaObject {
	return newQMetaObject(QBuffer_MetaObject(this.h))
}

func (this *QBuffer) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QBuffer_Metacast(this.h, param1_Cstring))
}

func QBuffer_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QBuffer_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QBuffer) Buffer() []byte {
	var _bytearray struct_miqt_string = QBuffer_Buffer(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QBuffer) Buffer2() []byte {
	var _bytearray struct_miqt_string = QBuffer_Buffer2(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QBuffer) SetData(data []byte) {
	data_alias := struct_miqt_string{}
	data_alias.data = (char)(unsafe.Pointer(&data[0]))
	data_alias.len = size_t(len(data))
	QBuffer_SetData(this.h, data_alias)
}

func (this *QBuffer) SetData2(data string, lenVal int64) {
	data_Cstring := CString(data)
	defer free(unsafe.Pointer(data_Cstring))
	QBuffer_SetData2(this.h, data_Cstring, (ptrdiff_t)(lenVal))
}

func (this *QBuffer) Data() []byte {
	var _bytearray struct_miqt_string = QBuffer_Data(this.h)
	_ret := GoBytes(unsafe.Pointer(_bytearray.data), int(int64(_bytearray.len)))
	free(unsafe.Pointer(_bytearray.data))
	return _ret
}

func (this *QBuffer) Open(openMode OpenMode) bool {
	return (bool)(QBuffer_Open(this.h, openMode))
}

func (this *QBuffer) Close() {
	QBuffer_Close(this.h)
}

func (this *QBuffer) Size() int64 {
	return (int64)(QBuffer_Size(this.h))
}

func (this *QBuffer) Pos() int64 {
	return (int64)(QBuffer_Pos(this.h))
}

func (this *QBuffer) Seek(off int64) bool {
	return (bool)(QBuffer_Seek(this.h, (longlong)(off)))
}

func (this *QBuffer) AtEnd() bool {
	return (bool)(QBuffer_AtEnd(this.h))
}

func (this *QBuffer) CanReadLine() bool {
	return (bool)(QBuffer_CanReadLine(this.h))
}

func QBuffer_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QBuffer_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QBuffer_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QBuffer_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QBuffer) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QBuffer_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QBuffer) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QBuffer_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QBuffer_MetaObject
func miqt_exec_callback_QBuffer_MetaObject(self QBuffer, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QBuffer{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QBuffer) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QBuffer_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QBuffer) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QBuffer_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QBuffer_Metacast
func miqt_exec_callback_QBuffer_Metacast(self QBuffer, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QBuffer{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qbuffer.cpp
// +build ignore

#include <QBuffer>
#include <QByteArray>
#include <QIODevice>
#include <QIODeviceBase>
#include <QMetaMethod>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qbuffer.h>
#include "gen_qbuffer.h"
class MiqtVirtualQBuffer : public virtual QBuffer {
public:
MiqtVirtualQBuffer(): QBuffer() {};
MiqtVirtualQBuffer(QObject* parent): QBuffer(parent) {};

virtual ~MiqtVirtualQBuffer() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QBuffer::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QBuffer_MetaObject(const_cast<MiqtVirtualQBuffer*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QBuffer::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QBuffer::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QBuffer_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QBuffer::qt_metacast(param1);

	}
};
QBuffer* QBuffer_new() {
return new MiqtVirtualQBuffer();
}
QBuffer* QBuffer_new2(QObject* parent) {
return new MiqtVirtualQBuffer(parent);
}
void QBuffer_virtbase(QBuffer* src
, QIODevice** outptr_QIODevice
) {
*outptr_QIODevice = static_cast<QIODevice*>(src);
}
QMetaObject* QBuffer_MetaObject(const QBuffer* self) {
	return (QMetaObject*) self->metaObject();
}
void* QBuffer_Metacast(QBuffer* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QBuffer_Tr(const char* s) {
	QString _ret = QBuffer::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QBuffer_Buffer(QBuffer* self) {
	QByteArray _qb = self->buffer();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}
struct miqt_string QBuffer_Buffer2(const QBuffer* self) {
	const QByteArray _qb = self->buffer();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}
void QBuffer_SetData(QBuffer* self, struct miqt_string data) {
	QByteArray data_QByteArray(data.data, data.len);
	self->setData(data_QByteArray);
}
void QBuffer_SetData2(QBuffer* self, const char* data, ptrdiff_t lenVal) {
	self->setData(data, (qsizetype)(lenVal));
}
struct miqt_string QBuffer_Data(const QBuffer* self) {
	const QByteArray _qb = self->data();
	struct miqt_string _ms;
	_ms.len = _qb.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _qb.data(), _ms.len);
	return _ms;
}
bool QBuffer_Open(QBuffer* self, OpenMode openMode) {
	return self->open(openMode);
}
void QBuffer_Close(QBuffer* self) {
	self->close();
}
long long QBuffer_Size(const QBuffer* self) {
	qint64 _ret = self->size();
	return static_cast<long long>(_ret);
}
long long QBuffer_Pos(const QBuffer* self) {
	qint64 _ret = self->pos();
	return static_cast<long long>(_ret);
}
bool QBuffer_Seek(QBuffer* self, long long off) {
	return self->seek(static_cast<qint64>(off));
}
bool QBuffer_AtEnd(const QBuffer* self) {
	return self->atEnd();
}
bool QBuffer_CanReadLine(const QBuffer* self) {
	return self->canReadLine();
}
struct miqt_string QBuffer_Tr2(const char* s, const char* c) {
	QString _ret = QBuffer::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QBuffer_Tr3(const char* s, const char* c, int n) {
	QString _ret = QBuffer::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QBuffer_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQBuffer*>( (QBuffer*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QBuffer_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQBuffer*)(self) )->virtualbase_MetaObject();
}
void QBuffer_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQBuffer*>( (QBuffer*)(self) )->handle__Metacast = slot;
}
void* QBuffer_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQBuffer*)(self) )->virtualbase_Metacast(param1);
}
void QBuffer_Delete(QBuffer* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQBuffer*>( self );
	} else {
		delete self;
	}
}
